### PR TITLE
update QuicError

### LINKED ISSUE #87259: [API Proposal]: udate QuicError values
### Background and motivation

This is primarily driven by #75115 and https://github.com/dotnet/runtime/issues/82262.

As noted in #75115, `AcceptConnectionAsync` can throw various exceptions and it is not obvious how they should be handled. For one, Quic also surfaces exceptions from user callback directly and that may make then indistinguishable from exceptions thrown by Quic itself - ODE may be particularly trick as leaked ODE exception from the callback may stop the server as it may look like whole listener was disposed. To make that more obvious, this issue is proposing to add new enum value and wrap anything from user callbacks in separate `QuicException`. #75184 originally proposed new enum specific to accept. However, after some discussion the user callback remains only one problematic and it is also conceptually valid to 'ConnectAsync`.

The second part of this proposal is based on API review board feedback on #76435. We wanted to add `QuicError.AddressNotAvailable` but the feedback was that is looks just like `SocketError`. As part of #82262  we agreed to throw `SocketException` for conditions that are not specific to Quic protocol and are caused by network conditions or environment configuration. That make several existing `QuicError` values obsolete and they would be never used. The proposal is to remove them - this is breaking change but we feel it is best long term option given that Quic still has preview flag. 
   

### API Proposal

```diff
namespace System.Net.Quic
{

    public enum QuicError
    {
        // existing members omitted

+       /// <summary>
+       /// An error occurred in user provided callback.
+       /// </summary>
+      UserCallbackError,
    }
}
```

```diff
    public enum QuicError
    {
        // existing members omitted

-      AddressInUse,
-      InvalidAddress,
-      HostUnreachable, // ICMP error
    }
```

### API Usage

from #75115 Kestrel fragment

```csharp
while (true)
{
    try
    {
        var connection = await _listener.AcceptConnectionAsync();
        return connection;
    }
    catch (ObjectDisposedException)
    {
        return null;
    }
    catch (QuicException ex)
    {
        // retry 
        log($"Quic failure {ex.Message} {ex.ErrorCode}");
    }
    catch (ArgumentException)
    {
        // retry
    }
}
```

### Alternative Designs

https://github.com/dotnet/runtime/issues/75184 originally proposed to add `QuicError.AcceptConnectionFailed`.

For the removal, we may mark the extra values as obsolete or we can keep them unused.

### Risks

The proposed removal is breaking change. I tried to search but I did not find code that would explicitly handle the transport error codes.

### PR DESCRIPTION
**Pull request by wfurt**
fixes #87259
contributes to https://github.com/dotnet/runtime/issues/75115
contributes to #87262

### CODE REVIEW

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicError.cs [hunk 1]
```diff
@@ -77,5 +62,15 @@ public enum QuicError
...
+
+        /// <summary>
+        /// Operation failed because peer transport error occurred.
+        /// </summary>
+        TransportError,
+
+        /// <summary>
+        /// An error occurred in user provided callback.
+        /// </summary>
+       UserCallbackError,
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/QuicError.cs:74]:
```suggestion
        UserCallbackError,
```

File: src/libraries/System.Net.Quic/ref/System.Net.Quic.cs [hunk 2]
```diff
@@ -51,16 +51,15 @@ public enum QuicError
...
         ConnectionTimeout = 6,
-        HostUnreachable = 7,
         ConnectionRefused = 8,
         VersionNegotiationError = 9,
         ConnectionIdle = 10,
         ProtocolError = 11,
         OperationAborted = 12,
         AlpnInUse = 13,
+        TransportError = 14,
+        UserCallbackError = 15,
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/ref/System.Net.Quic.cs:62]: It was approved just as `CallbackError`: https://github.com/dotnet/runtime/issues/87259#issuecomment-1603037411

**Review by ManickaP — APPROVED** [review]
LGTM modulo comments, thank you!

File: src/libraries/System.Net.Quic/ref/System.Net.Quic.cs (hunk 2 above)
Comment (wfurt) [reply src/libraries/System.Net.Quic/ref/System.Net.Quic.cs]: good catch. updated.

**Pull request merged**